### PR TITLE
Consensus: block batch prepper

### PR DESCRIPTION
**Pull request by Boog900**
This significantly speeds up performance compared to just prepping all blocks individually.

### CODE REVIEW

File: consensus/src/block.rs [hunk 1]
```diff
@@ -7,24 +7,82 @@ use std::{
...
+    pub block_hash: [u8; 32],
+    /// The height of the block.
+    pub height: u64,
+
+    /// The weight of the blocks miner transaction.
+    pub miner_tx_weight: usize,
+}
+
+impl PrePreparedBlockExPOW {
+    pub fn new(block: Block) -> Result<PrePreparedBlockExPOW, ConsensusError> {
```
Comment (hinto-janai) [inline consensus/src/block.rs:62]:
```suggestion
    /// Prepare a new block.
    ///
    /// # Errors
    /// This errors if either the `block`'s:
    /// - Hard-fork values are invalid
    /// - Miner transaction is missing
    pub fn new(block: Block) -> Result<PrePreparedBlockExPOW, ConsensusError> {
```

File: consensus/src/block.rs [hunk 2]
```diff
@@ -82,6 +140,32 @@ impl PrePreparedBlock {
             block,
         })
     }
+
+    /// Creates a new [`PrePreparedBlock`] from a [`PrePreparedBlockExPOW`].
+    ///
+    /// The randomX VM must be Some if RX is needed or this will panic.
+    /// The randomX VM must also be initialised with the correct seed.
+    fn new_prepped<R: RandomX>(
```
Comment (hinto-janai) [inline consensus/src/block.rs:148]:
```suggestion
    /// Creates a new [`PrePreparedBlock`] from a [`PrePreparedBlockExPOW`].
    ///
    /// # Errors
    /// This function will error if `randomx_vm is
    /// not initialised with the correct seed.
    ///
    /// # Panics
    /// This function will panic if `randomx_vm` is
    /// [`None`] even though RandomX is needed.
    fn new_prepped<R: RandomX>(
```

File: consensus/src/block.rs [hunk 3]
```diff
@@ -91,12 +175,28 @@ pub enum VerifyBlockRequest {
...
+        /// The already prepared block.
+        block: PrePreparedBlock,
+        /// The full list of transactions for this block, in order given in `block`.
+        txs: Vec<Arc<TransactionVerificationData>>,
+    },
+    /// Batch prepares a list of blocks and transactions for verification.
+    MainChainBatchPrepareBlocks {
+        /// The list of blocks.
+        blocks: Vec<(Block, Vec<Transaction>)>,
+    },
```
Comment (hinto-janai) [inline consensus/src/block.rs:189]:
```suggestion
    MainChainPrepped {
        /// The already prepared block.
        block: PrePreparedBlock,
        /// The full list of transactions for this block, in the same order as `block`'s.
        txs: Vec<Arc<TransactionVerificationData>>,
    },
    /// Batch prepares a list of blocks and transactions for verification.
    MainChainBatchPrepareBlocks {
        /// The list of blocks and their transactions (in the same order as the block's).
        blocks: Vec<(Block, Vec<Transaction>)>,
    },
```

File: consensus/src/block.rs [hunk 4]
```diff
@@ -178,17 +278,185 @@ where
...
+                        .await
+                }
             }
         }
         .boxed()
     }
 }
 
+/// Batch prepares a list of blocks for verification.
+#[instrument(level="debug", name="batch_prep_blocks" skip_all, fields(amt=blocks.len()))]
```
Comment (hinto-janai) [inline consensus/src/block.rs:295]:
```suggestion
/// Batch prepares a list of blocks for verification.
#[instrument(level = "debug", name = "batch_prep_blocks", skip_all, fields(amt = blocks.len()))]
```

File: consensus/src/block.rs [hunk 5]
```diff
@@ -178,17 +278,185 @@ where
...
+            Err(ConsensusError::Block(BlockError::PreviousIDIncorrect))?;
+        }
+
+        // Cache any potential RX VM seeds as we may need them for future blocks in the batch.
+        if is_randomx_seed_height(window[0].height) {
+            new_rx_vm = Some((window[0].height, window[0].block_hash));
+        }
+
+        timestamps_hfs.push((window[0].block.header.timestamp, window[0].hf_version))
+    }
```
Comment (hinto-janai) [inline consensus/src/block.rs:341]: Only need to `[]` once.
```suggestion
    for window in blocks.windows(2) {
        let block_1 = &window[0];
        let block_2 = &window[1];
        let block_1_hash = block_1.block_hash;
        let block_1_height = block_1.height;
        let block_2_height = block_2.height;

        if block_1_hash != block_2.block.header.previous || block_1_height != block_2_height - 1 {
            tracing::debug!("Blocks do not follow each other, verification failed.");
            Err(ConsensusError::Block(BlockError::PreviousIDIncorrect))?;
        }

        // Cache any potential RX VM seeds as we may need them for future blocks in the batch.
        if is_randomx_seed_height(block_1_height) {
            new_rx_vm = Some((block_1_height, block_1_hash));
        }

        timestamps_hfs.push((block_1.block.header.timestamp, block_1.hf_version))
    }
```

File: consensus/src/block.rs [hunk 6]
```diff
@@ -178,17 +278,185 @@ where
...
+
+    if context.chain_height != blocks[0].height {
+        tracing::debug!("Blocks do not follow main chain, verification failed.");
+
+        Err(ConsensusError::Block(BlockError::MinerTxError(
+            MinerTxError::InputsHeightIncorrect,
+        )))?;
+    }
+
+    if context.top_hash != blocks[0].block.header.previous {
```
Comment (hinto-janai) [inline consensus/src/block.rs:379]: `[]` once.
```suggestion
    let block_0 = &blocks[0];

    if context.chain_height != block_0.height {
        tracing::debug!("Blocks do not follow main chain, verification failed.");

        Err(ConsensusError::Block(BlockError::MinerTxError(
            MinerTxError::InputsHeightIncorrect,
        )))?;
    }

    if context.top_hash != block_0.block.header.previous {
```

File: consensus/src/block.rs [hunk 7]
```diff
@@ -201,63 +469,130 @@ where
...
+            .oneshot(BlockChainContextRequest::GetContext)
+            .await
+            .map_err(Into::<ExtendedConsensusError>::into)?
+        else {
+            panic!("Context service returned wrong response!");
+        };
+
+        let context = checked_context.unchecked_blockchain_context().clone();
+
+        tracing::debug!("got blockchain context: {:?}", context);
```
Comment (hinto-janai) [inline consensus/src/block.rs:561]:
```suggestion
        tracing::debug!("got blockchain context: {context:?}");
```

File: consensus/src/block.rs (hunk 5 above)
Comment (Boog900) [reply consensus/src/block.rs:351]: I really think the compiler would take care of this + I'd rather the readability

File: consensus/src/block.rs (hunk 6 above)
Comment (Boog900) [reply consensus/src/block.rs:389]: same here

File: consensus/src/block.rs [hunk 8]
```diff
@@ -7,24 +7,82 @@ use std::{
...
+        rx_vms::RandomXVM, BlockChainContextRequest, BlockChainContextResponse,
+        RawBlockChainContext,
+    },
     transactions::{TransactionVerificationData, VerifyTxRequest, VerifyTxResponse},
     Database, ExtendedConsensusError,
 };
 
+/// A pre-prepared block with all data needed to verify it, except the block's proof of work.
+#[derive(Debug)]
+pub struct PrePreparedBlockExPOW {
```
[the remark this answers is not in the document]
Comment (Boog900) [reply consensus/src/block.rs:41]: I'll keep it in mind, however I think it would make this name even more ugly

**Review by SyntheticBird45 — APPROVED** [review]

File: consensus/src/block.rs (hunk 5 above)
Comment (hinto-janai) [reply consensus/src/block.rs:351]: It's less about optimization and more about identifier safety, e.g. the following requires 1 ident modification when refactoring:
```rust
let x = &y[0];
let _ = x.a;
let _ = x.b;
let _ = x.c;
let _ = x.d;
```
but this requires 5 line changes:
```rust
let _ = &y[0];
let _ = &y[0].a;
let _ = &y[0].b;
let _ = &y[0].c;
let _ = &y[0].d;
```
It's harder to mess up when making changes in the 1st example.

File: consensus/src/block.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Boog900) [reply consensus/src/block.rs:41]: Fair, what would this even be: `PrepreparedBlockExPow`, `PrepreparedBlockExpow` or `PrePreparedBlockExpow`.

I think it woulds be `PrepreparedBlockExPow`, right?

[the remark this answers is not in the document]
Comment (hinto-janai) [reply consensus/src/block.rs:41]: I think `PrepreparedBlockExPow`, could also just be `PreparedBlockExPow`.

File: consensus/src/block.rs (hunk 5 above)
Comment (SyntheticBird45) [reply consensus/src/block.rs:351]: I prefer the original syntax, but your explanation make sense. I could say its *skill issue* from the dev to mess up this but its not really serious from a codebase perspective. Therefore I agree with your suggestion

Comment (Boog900) [reply consensus/src/block.rs]: I don't know I'm still pretty skeptical, I partially did the change though

File: consensus/src/block.rs (hunk 6 above)
Comment (Boog900) [reply consensus/src/block.rs:392]: I don't think it would help here though

**Review by hinto-janai — APPROVED** [review]

**Pull request merged**